chore: add migration guide for 0.7.x (#1238)
eclipse-tractusx/tractusx-edc: docs/migration/Version_0.5.x_0.7.x.md

# Migration Guide `0.5.x/0.6.0 -> 0.7.x`

<!-- TOC -->

* [Migration Guide `0.5.x/0.6.0 -> 0.7.x`](#migration-guide-05x060---07x)
    * [1. Dataplane Signaling (DPS)](#1-dataplane-signaling-dps)
        * [1.1 Overview](#11-overview)

>         * [1.2  Automatic token renewal](#12--automatic-token-renewal)

Reviewer: Typo or rather one space too much.

        * [2.1 Example of dynamic callback](#21-example-of-dynamic-callback)
        * [2.2 Example of static callbacks configuration](#22-example-of-static-callbacks-configuration)
        * [2.3 References](#23-references)
    * [3. Identity And Trust Protocols (IATP)](#3-identity-and-trust-protocols-iatp)
        * [3.1 Preconditions](#31-preconditions)
        * [3.2 Configuration values](#32-configuration-values)
        * [3.3 DIDs and BPNs](#33-dids-and-bpns)
        * [3.4 References](#34-references)
    * [4. BPN/DID Resolution Service](#4-bpndid-resolution-service)
        * [4.1 Config values](#41-config-values)
        * [4.2 References](#42-references)
    * [5. EDC Management API](#5-edc-management-api)
        * [5.1 Catalog](#51-catalog)
        * [5.2 Transfer Process](#52-transfer-process)
        * [5.3 EDRs](#53-edrs)
    * [6. Other changes](#6-other-changes)

<!-- TOC -->

## 1. Dataplane Signaling (DPS)

### 1.1 Overview

Dataplane Signaling effectively replaces the previous data plane public API. It is a standardized way for the control
plane to communicate with the data plane, e.g. when suspending or resuming data transfers.

> New features like token renewal build on to of it.

Reviewer: ```suggestion
New features like token renewal build on top of it.
```

> This replaces the Control API, the Public API and the EDR callbacks

### 1.2  Automatic token renewal

> when a `TransferProcess` enters the `STARTED` state, the consumer connector receives the Endpoint Data Reference (EDR)

Reviewer: ```suggestion
When a `TransferProcess` enters the `STARTED` state, the consumer connector receives the Endpoint Data Reference (EDR)
```

that contains the URL to the public API, an access token, a refresh token and endpoint where the token can be refreshed.

EDRs are now cached automatically within the consumer connector's control plane, and the [EDR API](#2-new-edr-api) will
auto-refresh the token if it is nearing expiry.

### 1.3 New data format for `DataAddress`

In order for Dataplane Signaling to become aligned with the Dataspace Protocol (DSP), `DataAddress` objects now conform
to
the `dspace` [data format](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling-token-handling.md#2-updates-to-thedataaddress-format).

### 1.4 Dataplane Signaling API

> this is a new REST api that replaces the old Control API for a more consolidated and streamlined communication between a

Reviewer: ```suggestion
This is a new REST api that replaces the old Control API for a more consolidated and streamlined communication between a
```

### 1.6 Further references

- Dataplane Signaling
  [documentation](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling.md)
  and [token
  handling](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling-token-handling.md)
- [Tractus-X Specification for Token
  Renewal](https://github.com/eclipse-tractusx/tractusx-profiles/blob/main/tx/refresh/refresh.token.grant.profile.md)
- [Tractus-X EDC Implementation
  Documentation](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/docs/development/dataplane-signaling/tx-signaling.extensions.md)

## 2. Changes to EDR handling and APIs

The setting `backendService.httpProxyTokenReceiverUrl`, which configured an EDR receiver backend using the upstream
EDC [extension](https://github.com/eclipse-edc/Connector/tree/main/extensions/control-plane/transfer/transfer-pull-http-dynamic-receiver)
is no longer available as well as the extension is not shipped in the Tractus-X EDC 0.7.x distributions.

> The main reason is that EDC/Tractus-X-Edc switched to Dataplane Signaling for handling transfers and the EDR format

Reviewer: ```suggestion
The main reason is that EDC/Tractus-X-EDC switched to Dataplane Signaling for handling transfers and the EDR format
```
Nitpicking here since it irks me to see EDC written differently in the same sentence.

specified in
DPS [spec](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling-token-handling.md)
was not handled properly by the receiver extension.

Now all the EDRs are automatically stored by the consumer's EDR cache and made available via the new

> EDRs API. If users still need to receive the EDR on their backend and hand this manually, EDC can handle this

Reviewer: ```suggestion
EDRs API. If users still need to receive the EDR on their backend and want to process this manually, EDC can handle this
```
Typo in handle and made an alternative suggestion.

- `bdrs-server-memory`: variant, that uses pure in-mem storage, useful for testing

To see how BDRS could be used in testing/dev etc. check out
this [test class](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/edc-extensions/bdrs-client/src/test/java/org/eclipse/tractusx/edc/identity/mapper/BdrsClientImplComponentTest.java)

> Note that consuming the Directory API with tools like Postman or Insomnia is possible, but not straight forward and it
> is out-of-scope for this document.

## 5. EDC Management API

### 5.1 Catalog

- `counterPartyId` is a **required** property for the catalog request in Tractus-X EDC. It's not enforced currently (
  backward compatibility) but if not provided, the IATP flows won't work and thus the catalog request will fail.

### 5.2 Transfer Process

- `transferType`  is a needed property for the transfer request to be working in tx-edc. `transferTypes` are the
  supported transfer format in tx-edc. They return as part of the Catalog response as part of distribution format for an
  asset. In TX-EDC we currently supports `HttpData-PULL`, `AzureStorage-PUSH` and `AmazonS3-PUSH`. The `transferType`

>   i's not mandatory (backward compatibility) but if not provided the transfer request will fail.

Reviewer: ```suggestion
  is not mandatory (backward compatibility) but if not provided the transfer request will fail.
```

### 5.3 EDRs

The 0.5.4 EDRs API was removed due the adoption of DPS (data plane signaling) and reborn under a different shape
in `/v2/edrs`

The main difference between 0.5.4 and 0.7.0 is how the token refresh is handled. In 0.5.x when an EDR was about to

> expire or the users had to manually trigger another transfer process, or by using the EDR API which did that

Reviewer: ```suggestion
expire, the users had to manually trigger another transfer process, or use the EDR API which did that
```
I think this wants to say how the token was refreshed previously but the sentence seems jumbled.

  asset. In TX-EDC we currently supports `HttpData-PULL`, `AzureStorage-PUSH` and `AmazonS3-PUSH`. The `transferType`
  i's not mandatory (backward compatibility) but if not provided the transfer request will fail.

### 5.3 EDRs

The 0.5.4 EDRs API was removed due the adoption of DPS (data plane signaling) and reborn under a different shape
in `/v2/edrs`

The main difference between 0.5.4 and 0.7.0 is how the token refresh is handled. In 0.5.x when an EDR was about to
expire or the users had to manually trigger another transfer process, or by using the EDR API which did that
automatically. This approach had major drawbacks like proactively creating transfer process on EDR expiration. In 0.7.0
this changed radically by adopting DPS and overlaying on top of it
a [refresh](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/docs/development/dataplane-signaling/tx-signaling.extensions.md)
mechanism for EDRs.

For a detailed usage I would suggest
changing [this](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/docs/usage/management-api-walkthrough/07_edrs.md)

> section of the docs.

Reviewer 2: It's already been changed, we can point to the docs directly

### 1.6 Further references

- Dataplane Signaling
  [documentation](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling.md)
  and [token
  handling](https://github.com/eclipse-edc/Connector/blob/main/docs/developer/data-plane-signaling/data-plane-signaling-token-handling.md)
- [Tractus-X Specification for Token
  Renewal](https://github.com/eclipse-tractusx/tractusx-profiles/blob/main/tx/refresh/refresh.token.grant.profile.md)
- [Tractus-X EDC Implementation
  Documentation](https://github.com/eclipse-tractusx/tractusx-edc/blob/main/docs/development/dataplane-signaling/tx-signaling.extensions.md)

## 2. Changes to EDR handling and APIs

The setting `backendService.httpProxyTokenReceiverUrl`, which configured an EDR receiver backend using the upstream
EDC [extension](https://github.com/eclipse-edc/Connector/tree/main/extensions/control-plane/transfer/transfer-pull-http-dynamic-receiver)
is no longer available as well as the extension is not shipped in the Tractus-X EDC 0.7.x distributions.

> 

Reviewer 3: ```suggestion
Likewise, the option to dynamically register a consumer-side HTTP-callback via `receiverHttpEndpoint` in the POST /transferprocesses call was removed. Consumer apps should use the /v2/edrs API instead.

```

> Note that consuming the Directory API with tools like Postman or Insomnia is possible, but not straight forward and it
> is out-of-scope for this document.

## 5. EDC Management API

### 5.1 Catalog

- `counterPartyId` is a **required** property for the catalog request in Tractus-X EDC. It's not enforced currently (
  backward compatibility) but if not provided, the IATP flows won't work and thus the catalog request will fail.

### 5.2 Transfer Process

- `transferType`  is a needed property for the transfer request to be working in tx-edc. `transferTypes` are the
  supported transfer format in tx-edc. They return as part of the Catalog response as part of distribution format for an
  asset. In TX-EDC we currently supports `HttpData-PULL`, `AzureStorage-PUSH` and `AmazonS3-PUSH`. The `transferType`
  is not mandatory (backward compatibility) but if not provided the transfer request will fail.

### 5.3 EDRs

The 0.5.4 EDRs API was removed due the adoption of DPS (data plane signaling) and reborn under a different shape

> in `/v2/edrs`

Reviewer 3: Suggest to add:

"Additionally, the option to register a consumer-side HTTP-callback via `receiverHttpEndpoint` in the `POST /transferprocesses` call was removed. Consumer apps should use the `/v2/edrs` API instead."

Reviewer 2: Probably we can add this in the section  `## 2. Changes to EDR handling and APIs` where we mention the removal of the config `backendService.httpProxyTokenReceiverUrl`. We can add the removal of the `receiverHttpEndpoint` in transfer process as well

Author: do you mind adding a GitHub Suggestion to the PR? CTRL-G (or CMD-G on macs)...
